plug.vim

```diff
@@ -968,6 +973,377 @@ while 1 " Without TCO, Vim stack is bound to explode
+  def write(cls, msg):
+    if cls.ON == None:
+      cls.ON = int(vim.eval('get(g:, "plug_log_on", 0)'))
+      cls.LOGDIR = os.path.expanduser(vim.eval('get(g:, "plug_logs", "~/plug_logs")'))
+    if cls.ON:
+      if not os.path.exists(cls.LOGDIR):
+        os.makedirs(cls.LOGDIR)
+      cls._write(msg)
+  @classmethod
+  def _write(cls, msg):
+    name = thr.current_thread().name
+    fname = cls.LOGDIR + os.path.sep + name
+    with open(fname, 'ab') as flog:
+      ltime = datetime.datetime.now().strftime("%H:%M:%S.%f")
+      msg = '[{},{}] {}{}'.format(name, ltime, msg, '\n')
+      flog.write(msg)
+
+class Buffer(object):
+  """ Lock before calling write. """
+  def __init__(self, lock, num_plugs):
+    self.bar = ''
+    self.event = 'Updating' if vim.eval('s:update.pull') == '1' else 'Installing'
+    self.is_win = vim.eval('s:is_win') == '1'
+    self.lock = lock
+    self.maxy = int(vim.eval('winheight(".")'))
+    self.num_plugs = num_plugs
+
+  def where(self, name):
+    """ Helper function, given a name find first occurence in buffer lines. """
+    found, lnum = False, 0
+    matcher = re.compile('^[-+x*] {}:'.format(name))
+    for line in vim.current.buffer:
+      if matcher.search(line) != None:
+        found = True
+        break
+      lnum += 1
+
+    if not found:
+      lnum = -1
+    return lnum
+
+  def header(self):
+    """ Prints the header at the top. """
+    curbuf = vim.current.buffer
+    curbuf[0] = self.event + ' plugins ({}/{})'.format(len(self.bar), self.num_plugs)
+
+    num_spaces = self.num_plugs - len(self.bar)
+    curbuf[1] = '[{}{}]'.format(self.bar, num_spaces * ' ')
+
+    vim.command('normal! 2G')
+    if not self.is_win:
+      vim.command('redraw')
+
+  def write(self, *args, **kwargs):
+    """ Simply synchronizes calls to write. """
+    with self.lock:
+      self._write(*args, **kwargs)
+
+  def _write(self, action, name, lines):
+    """ Main logging function, updates based on threads calling. """
+    if action == Action.ERROR:
+      self.bar += 'x'
+      vim.command("call add(s:update.errors, '{}')".format(name))
+    elif action == Action.DONE:
+      self.bar += '='
+
+    first, rest = lines[0], lines[1:]
+    msg = ['{} {}: {}'.format(action, name, first)]
+    padded_rest = ['    ' + line for line in rest]
+    msg.extend(padded_rest)
+
+    curbuf = vim.current.buffer
+    lnum = self.where(name)
+    if lnum != -1: # Found matching line num
+      del curbuf[lnum]
+      if lnum > self.maxy and action in {Action.INSTALL, Action.UPDATE}:
+        lnum = 3
+    else:
+      lnum = 3
+    curbuf.append(msg, lnum)
+
+    self.header()
+
+class Command(object):
+  def __init__(self, cmd, cmd_dir=None, timeout=60, ntries=3, cb=None, clean=None):
+    self.cmd = cmd
+    self.cmd_dir = cmd_dir
+    self.timeout = timeout
+    self.ntries = ntries
+
+    self.clean = clean
+    if cb:
+      self.callback = cb
+    else:
+      def dummy(msg): pass
+      self.callback = dummy
```

Comment:
Might as well use a `lambda` here:

```python
self.callback = lambda msg: None  # no-op
```

Writing `pass` is basically just another way of writing `return None` because all functions implicitly returns `None` if not otherwise specified:

```python
>>> def dummy(msg): pass
...
>>> foo = dummy('bar')
>>> type(foo)
<type 'NoneType'>
>>> print foo
None
```